br/pkg/lightning/backend/external/file.go

```diff
@@ -0,0 +1,181 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package external
+
+import (
+	"context"
+	"encoding/binary"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/pingcap/tidb/br/pkg/storage"
+)
+
+// KeyValueStore stores key-value pairs and maintains the range properties.
+type KeyValueStore struct {
+	dataWriter storage.ExternalFileWriter
+
+	rc        *rangePropertiesCollector
+	ctx       context.Context
+	offset    uint64
+	u64Buffer []byte
+}
+
+// NewKeyValueStore creates a new KeyValueStore. The data will be written to the
+// given dataWriter and range properties will be maintained in the given
+// rangePropertiesCollector.
+func NewKeyValueStore(
+	ctx context.Context,
+	dataWriter storage.ExternalFileWriter,
+	rangePropertiesCollector *rangePropertiesCollector,
+) (*KeyValueStore, error) {
+	kvStore := &KeyValueStore{
+		dataWriter: dataWriter,
+		ctx:        ctx,
+		rc:         rangePropertiesCollector,
+		u64Buffer:  make([]byte, 8),
+	}
+	return kvStore, nil
+}
+
+// AddKeyValue saves a key-value pair to the KeyValueStore. If the accumulated
+// size or key count exceeds the given distance, a new range property will be
+// appended to the rangePropertiesCollector with current status.
+// Caller should guarantee `writerID` and `seq` are not changed.
```

Comment:
maybe we can specify the writerID and seq when `NewKeyValueStore`? @tangenta @wjhuang2016